### PR TITLE
Refactor SessionController

### PR DESCRIPTION
**Pull request by BeeHiveJava**

### CODE REVIEW

File: Soccar [Game-Server]/src/nl/soccar/gameserver/SessionController.java [hunk 1]
```diff
@@ -61,128 +58,88 @@ public boolean createSession(String name, String password, String hostName, int
...
         synchronized (sessions) {
-            Optional<Session> optional = sessions.stream().filter(s -> s.getRoom().getName().equals(name)).findAny();
-            if (optional.isPresent()) {
-                sessions.remove(optional.get());
+            SessionWrapper wrapper = sessions.remove(name);
+            if (wrapper == null) {
+                return;
             }
+
+            // TODO remove all players, or something...
```
Comment (ghost) [inline Soccar [Game-Server]/src/nl/soccar/gameserver/SessionController.java:77]: INFO Complete the task associated to this TODO comment. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1135)

File: Soccar [Game-Server]/src/nl/soccar/gameserver/SessionController.java [hunk 2]
```diff
@@ -25,7 +23,6 @@
 import nl.soccar.socnet.connection.Connection;
```
Comment (TehMessiah) [inline Soccar [Game-Server]/src/nl/soccar/gameserver/SessionController.java]: MINOR Remove this unused import 'nl.soccar.socnet.connection.Connection'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import nl.soccar.socnet.Node;
+import nl.soccar.socnet.connection.Connection;
+
+public final class SessionWrapper {
+
+    private final Node server;
+    private final SessionController controller;
+    private final Session session;
+
+    public SessionWrapper(Node server, SessionController controller, Session session) {
```
Comment (ghost) [inline Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java:18]: MINOR Document this public constructor. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        this.server = server;
+        this.controller = controller;
+        this.session = session;
+    }
+
+    public JoinSessionMessage.Status join(Player player, String password) {
```
Comment (ghost) [inline Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java:24]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+        PlayerLeftSessionMessage m = new PlayerLeftSessionMessage(player.getUsername(), team.getTeamColour());
+        room.getAllPlayers().stream().map(server::getConnectionFromPlayer).forEach(c -> c.send(m));
+
+        if (room.getCapacity() <= 0) {
+            controller.destroySession(session);
+        }
+    }
+
+    public void startGame() {
```
Comment (ghost) [inline Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java:90]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    private void sendJoinedSessionNotification(Player player, TeamColour team) {
+        Room room = session.getRoom();
+
+        // Send notification to all players that the connecting player joined
+        PlayerJoinedSessionMessage m = new PlayerJoinedSessionMessage(player, team);
+        room.getAllPlayers().stream().map(server::getConnectionFromPlayer).forEach(c -> c.send(m));
+    }
+
+    public void leaveSession(Player player) {
```
Comment (ghost) [inline Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java:75]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import nl.soccar.gameserver.message.JoinSessionMessage;
+import nl.soccar.gameserver.message.PlayerJoinedSessionMessage;
+import nl.soccar.gameserver.message.PlayerLeftSessionMessage;
+import nl.soccar.gameserver.message.PlayerStartedGameMessage;
+import nl.soccar.library.*;
+import nl.soccar.library.enumeration.TeamColour;
+import nl.soccar.socnet.Node;
+import nl.soccar.socnet.connection.Connection;
+
+public final class SessionWrapper {
```
Comment (ghost) [inline Soccar [Game-Server]/src/nl/soccar/gameserver/SessionWrapper.java:12]: MINOR Document this public class. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Game-Server]/src/nl/soccar/gameserver/rmi/GameServerController.java [hunk 8]
```diff
@@ -92,28 +91,36 @@ public GameServerController() {
...
             mainServerForGameServer.sessionCreated(gameServerForMainServer, session);
-            
+
         } catch (UnknownHostException | RemoteException e) {
             LOGGER.log(Level.SEVERE, "An error occurred while reporting a created session.", e);
         }
     }
-    
+
+    public void sessionDestroyed(String roomName) {
```
Comment (ghost) [inline Soccar [Game-Server]/src/nl/soccar/gameserver/rmi/GameServerController.java:118]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by TehMessiah** [comment]
SonarQube analysis reported 12 issues
* MINOR 11 minor
* INFO 1 info

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [SessionController.java#L10](https://github.com/PTS3-S34A/Game-Server/blob/b96127dee96cd62c99e14736e0125770af6f253b/Soccar [Game-Server]/src/nl/soccar/gameserver/SessionController.java#L10): Remove this unused import 'nl.soccar.gameserver.message.PlayerJoinedSessionMessage'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [SessionController.java#L11](https://github.com/PTS3-S34A/Game-Server/blob/b96127dee96cd62c99e14736e0125770af6f253b/Soccar [Game-Server]/src/nl/soccar/gameserver/SessionController.java#L11): Remove this unused import 'nl.soccar.gameserver.message.PlayerLeftSessionMessage'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [SessionController.java#L12](https://github.com/PTS3-S34A/Game-Server/blob/b96127dee96cd62c99e14736e0125770af6f253b/Soccar [Game-Server]/src/nl/soccar/gameserver/SessionController.java#L12): Remove this unused import 'nl.soccar.gameserver.message.PlayerStartedGameMessage'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [SessionController.java#L20](https://github.com/PTS3-S34A/Game-Server/blob/b96127dee96cd62c99e14736e0125770af6f253b/Soccar [Game-Server]/src/nl/soccar/gameserver/SessionController.java#L20): Remove this unused import 'nl.soccar.library.enumeration.TeamColour'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by ghost** [comment]
SonarQube analysis reported 7 issues
* MINOR 6 minor
* INFO 1 info

Watch the comments in this conversation to review them.

**Review by Ranray — APPROVED** [review]

**Pull request merged**